readme.md
# React Spotlight

This component covers the page with an overlay, only showing around coordinates that you choose. You can change the coordinates and the circle will animate to the new ones. This could be used to dynamically introduce parts of a website to the user, or create more complex interactive experiences.

## Usage

Live demo: https://www.matislepik.com/react-spotlight/

Source code can be found in the `example/src/Demo.js`.

### Installation

You can install the component through NPM:

`yarn add react-spotlight` or `npm install react-spotlight --save`

Then import it in your code:

`import Spotlight from 'react-spotlight';`

### Basic example:
Renders a spotlight at `x: 350px` `y: 25px`

```
// At top of file
import Spotlight from 'react-spotlight';

// In render method
<Spotlight x={350} y={25} />
```

### Example with all props:
Renders a spotlight at `x: 50%` `y: 25%`. You can add text content inside the circle, but you'll have to style it yourself.
```
<Spotlight
  x={50}
  y={25}
  color="#d74793"
  radius={100}
  responsive
  usePercentage
  animSpeed={1000}
  borderColor="#ddd"
  borderWidth={5}>
  <div style={{
    position: 'absolute',
    left: '50%',
    top: '-50px',
    transform: 'translate(-50%, -100%)',
    whiteSpace: 'nowrap'
  }}>
    <h1>Text content</h1>
  </div>
</Spotlight>
```

### Animating
You can change the `x` / `y` / `color` / `borderColor` variables dynamically and the spotlight will animate to the new size / position / colors. Radius can also be animated, but this is laggy because it changes the CSS width/height values.

### PropTypes

|PropType (*type*) = default|Description|
|---|---|
|x (*number*) = 50|Center of the circle. Pixels or percentage, depending on `usePercentage` value|
|y (*number*) = 50|Center of the circle. Pixels or percentage, depending on `usePercentage` value|
|radius (*number*) = 100|Radius of the circle. Avoid changing this dynamically because css transitions on width/height are laggy.|
|color (*string*) = 'black'|Color of the overlay|
|usePercentage (*boolean*) = false|By default, the coordinates are interpreted as pixels. This causes them to be interpreted as percentage of the viewport.|
|animSpeed (*number*) = 500|Transition speed when changing coordinates. You can control this manually with CSS if you want - but be sure to change it for both the outer and inner container.|
|borderColor (*string*) = 'white'|Color of the circle's border|
|borderWidth (*number*) = 0|Width of the circle's border|
|responsive (*boolean*) = true|Certain values depend on cached viewport size for performance, so we listen to a (throttled) resize event to recalculate values. Turning this off might result in buggy behaviour if the screen is resized.|
|outerClass|Class name for the outer container|
|innerClass|Class name for the inner container|
|outerStyles|Styles for the outer container. Useful if you want to customize the visuals of the spotlight. Be careful - these can override styles that are necessary to make the visuals work.|
|innerStyles|Styles for the inner container|
|children|You can render children into the inner container, but you will have to position them yourself.|

## Development

The example directory doubles as a place to test out changes to the component. It is created using [create-react-app](https://github.com/facebookincubator/create-react-app). To use it:

- `npm run build`
- `cd example`
- `yarn start`

## Notes

- This component works with server rendering, however there will be an invalid checksum error.
- This is inspired by [Ahmad Moussawi's Hop library](https://ahmad-moussawi.github.io/hop/) for jQuery.


CHANGELOG.md
# Changelog

## [1.1.0] - 2017-07-12
### Added
- `outerStyles` and `innerStyles` properties


test/utils/dom.js
const jsdom = require('jsdom').jsdom;

global.document = jsdom('<!doctype html><html><body></body></html>');
global.window = global.document.defaultView;

propagateToGlobal(global.window);

function propagateToGlobal(window) {
  for (const key in window) {
    if (!window.hasOwnProperty(key)) continue;
    if (key in global) continue;

    global[key] = window[key];
  }
}


test/utils/setup.js
#!/usr/bin/env node
/* eslint-disable */
//  enable runtime transpilation to use ES6/7 in node
var fs = require('fs');

var babelrc = fs.readFileSync('./.babelrc');
var config;

try {
  config = JSON.parse(babelrc);
} catch (err) {
  console.error('==>     ERROR: Error parsing your .babelrc.');
  console.error(err);
}

require('babel-register')(config);


test/render/HighlightArea.spec.js
import expect from 'expect';
import Spotlight from '../../dist/index';
import React from 'react';
import ReactDOM from 'react-dom';
import TestUtils from 'react-dom/test-utils';

describe('Spotlight', () => {
  it('Should render without blowing up', () => {
    const component = TestUtils.renderIntoDocument(<Spotlight />);
    expect(component).toExist();
  });

  it('Should add correct classes and styles', () => {
    const component = TestUtils.renderIntoDocument(
      <Spotlight outerClass="outer" innerClass="inner" />
    );
    const outer = TestUtils.findRenderedDOMComponentWithClass(
      component,
      'outer'
    );
    const inner = outer.querySelector('.inner');

    expect(outer).toExist();
    expect(inner).toExist();

    expect(outer.style.left).toNotEqual('');
    expect(outer.style.top).toNotEqual('');
    expect(outer.style.transform).toNotEqual('');
    expect(outer.style.width).toNotEqual('');
    expect(outer.style.height).toNotEqual('');

    expect(inner.style.width).toNotEqual('');
    expect(inner.style.height).toNotEqual('');
  });

  it('Should correctly calculate border size based on screen diagonal and circle radius', () => {
    const originalWindowSize = { w: window.innerWidth, h: window.innerHeight };
    window.innerWidth = 1920;
    window.innerHeight = 1080;

    const component1 = TestUtils.renderIntoDocument(
      <Spotlight radius={150} outerClass="outer" innerClass="inner" />
    );
    const component2 = TestUtils.renderIntoDocument(
      <Spotlight radius={25} outerClass="outer" innerClass="inner" />
    );
    const inner1 = TestUtils.findRenderedDOMComponentWithClass(
      component1,
      'inner'
    );
    const inner2 = TestUtils.findRenderedDOMComponentWithClass(
      component2,
      'inner'
    );

    expect(inner1.style.borderWidth).toBe('2353px');
    expect(inner2.style.borderWidth).toBe('2228px');

    // Mock window size
    window.innerWidth = 800;
    window.innerHeight = 600;

    const component3 = TestUtils.renderIntoDocument(
      <Spotlight radius={150} outerClass="outer" innerClass="inner" />
    );
    const component4 = TestUtils.renderIntoDocument(
      <Spotlight radius={25} outerClass="outer" innerClass="inner" />
    );
    const inner3 = TestUtils.findRenderedDOMComponentWithClass(
      component3,
      'inner'
    );
    const inner4 = TestUtils.findRenderedDOMComponentWithClass(
      component4,
      'inner'
    );

    expect(inner3.style.borderWidth).toBe('1150px');
    expect(inner4.style.borderWidth).toBe('1025px');

    // Restore window size
    window.innerWidth = originalWindowSize.w;
    window.innerHeight = originalWindowSize.h;
  });

  it('Should pass through inner/outerStyles', () => {
    const component = TestUtils.renderIntoDocument(
      <Spotlight
        outerClass="outer"
        innerClass="inner"
        outerStyles={{ zIndex: 123 }}
        innerStyles={{ zIndex: 456 }}
      />
    );
    const outer = TestUtils.findRenderedDOMComponentWithClass(
      component,
      'outer'
    );
    const inner = outer.querySelector('.inner');

    expect(outer.style.zIndex).toBe('123');
    expect(inner.style.zIndex).toBe('456');
  });
});


src/index.js
import React from 'react';
import PropTypes from 'prop-types';

export default class Spotlight extends React.Component {
  constructor(props) {
    super(props);

    const isServerSide = typeof window === 'undefined';
    // If the user has server rendering, we assume they have a 4k screen (better aim high than low just in case).
    this.state = {
      windowWidth: isServerSide ? 3840 : window.innerWidth,
      windowHeight: isServerSide ? 2160 : window.innerHeight,
    };

    this.resizeThrottler = this.resizeThrottler.bind(this);
  }

  componentDidMount() {
    if (this.props.responsive)
      window.addEventListener('resize', this.resizeThrottler);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.resizeThrottler);
  }

  // https://developer.mozilla.org/en-US/docs/Web/Events/resize#setTimeout
  resizeThrottler() {
    // ignore resize events as long as an actualResizeHandler execution is in the queue
    if (!this.resizeTimeout) {
      this.resizeTimeout = setTimeout(() => {
        this.resizeTimeout = null;
        this.recalc();
      }, 66); // The actualResizeHandler will execute at a rate of 15fps
    }
  }

  recalc() {
    this.setState({
      windowWidth: window.innerWidth,
      windowHeight: window.innerHeight,
    });
  }

  getBorderSize() {
    // We find the diagonal of the screen and add the circle's radius to ensure
    // that the border covers the screen when the circle is just off screen.
    const diagonal = Math.ceil(
      Math.sqrt(
        Math.pow(this.state.windowWidth, 2) +
          Math.pow(this.state.windowHeight, 2)
      )
    );
    return diagonal + this.props.radius;
  }

  getX() {
    const pos = this.props.usePercentage
      ? this.state.windowWidth * this.props.x / 100
      : this.props.x;
    return `${pos}px`;
  }

  getY() {
    const pos = this.props.usePercentage
      ? this.state.windowHeight * this.props.y / 100
      : this.props.y;
    return `${pos}px`;
  }

  getInnerStyles() {
    const { animSpeed, radius, borderWidth, color, innerStyles } = this.props;
    const diameter = radius * 2 + borderWidth * 2;

    return {
      position: 'absolute',
      top: '50%',
      left: '50%',
      width: diameter,
      height: diameter,
      transform: 'translate(-50%, -50%)',
      border: `${this.getBorderSize()}px solid ${color}`,
      opacity: 0.9,
      borderRadius: '50%',
      boxSizing: 'content-box',
      transitionDuration: `${animSpeed}ms`,
      transitionProperty: 'all',
      ...innerStyles,
    };
  }

  getOuterStyles() {
    const {
      animSpeed,
      borderWidth,
      borderColor,
      radius,
      outerStyles,
    } = this.props;
    const diameter = radius * 2;

    return {
      position: 'fixed',
      left: -radius - borderWidth,
      top: -radius - borderWidth,
      width: diameter,
      height: diameter,
      borderRadius: '50%',
      zIndex: 999,
      boxSizing: 'content-box',
      border: `${borderWidth}px solid ${borderColor}`,
      transform: `translate(${this.getX()}, ${this.getY()})`,
      transitionDuration: `${animSpeed}ms`,
      transitionProperty: 'all',
      ...outerStyles,
    };
  }

  render() {
    return (
      <div
        style={this.getOuterStyles()}
        className={this.props.outerClass || ''}
      >
        <div
          style={this.getInnerStyles()}
          className={this.props.innerClass || ''}
        >
          {this.props.children}
        </div>
      </div>
    );
  }
}

Spotlight.propTypes = {
  color: PropTypes.string,
  x: PropTypes.number,
  y: PropTypes.number,
  radius: PropTypes.number,
  responsive: PropTypes.bool,
  outerClass: PropTypes.string,
  innerClass: PropTypes.string,
  outerStyles: PropTypes.object,
  innerStyles: PropTypes.object,
  borderWidth: PropTypes.number,
  borderColor: PropTypes.string,
  animSpeed: PropTypes.number,
  usePercentage: PropTypes.bool,
};

Spotlight.defaultProps = {
  color: 'black',
  x: 50,
  y: 50,
  radius: 100,
  responsive: true,
  usePercentage: false,
  borderWidth: 0,
  borderColor: 'white',
  animSpeed: 500,
  outerStyles: {},
  innerStyles: {},
};


example/src/App.js
import React, { Component } from 'react';
import Demo from './Demo';
import './App.css';

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      x: -50,
      y: -15,
      radius: 100
    };
  }

  render() {
    return (
      <div>
        <header className="site-header">
          <div className="maxwidth-wrapper">
            <h1>React Spotlight</h1>
            <p>Highlight an area on your page by dimming the rest of the content.</p>
            <Demo />
          </div>
        </header>
        <div className="github-bar">
          <div className="maxwidth-wrapper github-bar-content">
            <code className="snippet">npm install react-spotlight</code>
            <span>or</span>
            <a className="link-view-docs" style={{ zIndex: '999' }} href="https://github.com/MatisLepik/react-spotlight">View docs on GitHub</a>
          </div>
        </div>

        <main className="usage maxwidth-wrapper">
          <h2>Basic example</h2>
          <code className="code-block" dangerouslySetInnerHTML={{ __html: `
<span class="source js jsx"><span class="meta tag jsx"><span class="punctuation definition tag jsx">&lt;</span><span class="entity name tag open jsx"><span class="support class component open jsx">Spotlight</span></span><span class="JSXAttrs"> <span class="entity other attribute-name jsx">x</span><span class="keyword operator assignment jsx">=</span><span class="meta embedded expression js"><span class="punctuation section embedded begin jsx">{</span><span class="source js jsx"><span class="constant numeric js">350</span></span><span class="punctuation section embedded end jsx">}</span></span> <span class="entity other attribute-name jsx">y</span><span class="keyword operator assignment jsx">=</span><span class="meta embedded expression js"><span class="punctuation section embedded begin jsx">{</span><span class="source js jsx"><span class="constant numeric js">25</span></span><span class="punctuation section embedded end jsx">}</span></span> </span><span class="punctuation definition tag jsx">/&gt;</span></span></span>
          ` }} />

          <h2>With all props</h2>
          <code className="code-block" dangerouslySetInnerHTML={{ __html: `
            &lt;Spotlight<br />
              &nbsp;&nbsp;x={50}<br />
              &nbsp;&nbsp;y={25}<br />
              &nbsp;&nbsp;color=&quot;#d74793&quot;<br />
              &nbsp;&nbsp;radius={100}<br />
              &nbsp;&nbsp;responsive<br />
              &nbsp;&nbsp;usePercentage<br />
              &nbsp;&nbsp;animSpeed={1000}<br />
              &nbsp;&nbsp;borderColor=&quot;black&quot;<br />
              &nbsp;&nbsp;borderWidth={5} /&gt;
          ` }}></code>
        </main>
      </div>
    );
  }
}

export default App;


example/src/Demo.js
import React from 'react';
import Spotlight from '../../dist/index';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';

export default class Demo extends React.Component {

  constructor(props) {
    super(props);

    this.start = this.start.bind(this);
    this.state = {
      playing: false,
      x: 0,
      y: 0,
      radius: 75,
      color: '#d74793',
      borderColor: '#1976d2'
    };
  }

  start() {
    this.setStatePromise({ playing: true, ...this.getCoordinates('.btn-try') })
      .then(() => this.sleep(800))
      .then(() => this.setStatePromise({ radius: 200, ...this.getCoordinates('.code-block .source') }))
      .then(() => this.sleep(2000))
      .then(() => this.setStatePromise({ radius: 100, ...this.getCoordinates('.link-view-docs') }))
      .then(() => this.sleep(1500))
      .then(() => this.setStatePromise({ color: '#1976d2', borderColor: 'rgba(0, 50, 225, .5)', text: 'Hello world!' }))
      .then(() => this.sleep(2000))
      .then(() => this.setStatePromise({ playing: false }));
  }

  /* The following are some convenience methods to make chaining the animation more convenient */

  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  setStatePromise(state) {
    this.setState(state);
    return Promise.resolve();
  }

  getCoordinates(selector) {
    const domEl = document.querySelector(selector);
    if (!domEl) return {};

    const rect = document.querySelector(selector).getBoundingClientRect();
    return { x: rect.left + (rect.width / 2), y: rect.top + (rect.height / 2) };
  }

  render() {
    // Note: the transition group and wrapper div aren't necessary.
    // They're just here to make the demo feel smoother
    return (
      <div>
        <button onClick={this.start} className="btn btn-try">Try it out!</button>

        <ReactCSSTransitionGroup
          transitionName="fade"
          transitionEnterTimeout={500}
          transitionLeaveTimeout={300}>
          {this.state.playing && (
            <div key="1">
              <Spotlight
                x={this.state.x}
                y={this.state.y}
                radius={this.state.radius}
                color={this.state.color}
                outerStyles={{ zIndex: 0 }}
                responsive
                animSpeed={1000}
                borderColor={this.state.borderColor}
                borderWidth={10}>
                <div style={{
                  position: 'absolute',
                  left: '50%',
                  top: '-50px',
                  transform: 'translate(-50%, -100%)',
                  whiteSpace: 'nowrap'
                }}>
                  <h1 style={{ margin: 0 }}>{this.state.text}</h1>
                </div>
              </Spotlight>
            </div>
          )}
        </ReactCSSTransitionGroup>
      </div>
    );
  }
}
